programmer-skills/sorting-algorithms.md

```diff
@@ -0,0 +1,442 @@
+	if (array.length < 2) { return array; }
+	var middle = Math.floor(array.length/2);
+	var left_array = array.slice(0, middle);
+	var right_array = array.slice(middle);
+	
+	return merge(mergeSort(left_array), mergeSort(right_array));
+}
+```
+
+###Heapsort  
+
+It is an inplace sorting algorithm implemented using the heap datastructure. Thus, we first have to explore what is a heap and how to build one. 
+
+
+####The heap data structure
+
+The heap data structure is an array that can be interpreted as a complete binary tree.
+Each element of the array is represented as a node in the binary tree.
+Every new level is filled from left to right. Only the last level can have missing elements.
+When we use an array to represent a heap it has two important characteristics:
+* the length of the array, which is equal to the total number of elements.
+* the heapsize, the number of elements in the heap. (Note: during the heapsort algorithm the number of elements in the heap changes dynamically. )
+
+
+There are two types of heap:  
+1. Max heap: 	
+&emsp;&emsp;`array[parent(i)]` &ge; `array[i]` is true if `i` is not the root element.  
+2. Min heap:  
+&emsp;&emsp;`array[parent(i)]` &le; `array[i]` is true if `i` is not the root element.  
+For clarity we will always use a max heap in this note. We will refer to (1.) as heap property.
+
+
+Figure 8. explains the relation between an array and the corresponding heap:  
+![heap as an array](http://www.stoimen.com/blog/wp-content/uploads/2012/08/4.-Heap-as-an-Array.png)  
+*Fig 8. taken from [here](http://www.stoimen.com/blog/2012/08/07/computer-algorithms-heap-and-heapsort-data-structure/)*  
+ 
+The JavaScript code for finding a node's parent and child elements:  
+ ```javascript
+function parent(i){
+	return Math.ceil( i / 2)-1;
+}
+
+function leftChild(i){
+	return 2*i+1;
+}
+
+function rightChild(i){
+return 2*i+2;
+}
+```
+
+The functions in our implementation:   
+1. heapSort: we invoke it on the array to be sorted.  
+2. buildHeap: builds a heap from an array.  
+3. heapify: tests whether all nodes satisfy the heap property.  
+4. swap: swaps elements if they violate the heap property. 
+
+The heapsort algorithm:      
+1. Build a max/min heap from the array to be sorted, now  `heapsize = array.length`  
+2. While `heapsize > 1`  
+&emsp;Swap the first element (mininum or maximum value depending on your implementation) with last one,  
+&emsp; reduce the `heapsize` by one,  then heapify the heap.  
+3. The array is sorted.  
+
+####Building the heap  
+
+The items in the lowest tree level have no child nodes to compare against, thus they are considered sorted. Good for us! 
+We start from `Math.floor( array.length / 2) ` counting backwards. 
+
+![Random array to heap](http://www.stoimen.com/blog/wp-content/uploads/2012/08/7.-Random-Array-to-Heap.png)   
+*Fig 9. is taken from [here](http://www.stoimen.com/blog/2012/08/07/computer-algorithms-heap-and-heapsort-data-structure/)*   
+
+####Heapify
```

Comment:
If I understand this correctly, the heapify needs to be done several times recursively, i.e. once you've swapped you start from `Math.floor( array.length / 2) ` again? If that's the case it might be worth spelling it out?